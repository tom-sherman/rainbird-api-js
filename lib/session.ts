import { Fetch } from './fetch'
import { btoa } from 'isomorphic-base64'
import { Fact } from './fact'

export class Session {
  readonly context?: string
  readonly apiDomain: string
  private _id?: string
  private _kmId: string
  private apiKey: string

  constructor({ apiKey, context, apiDomain, kmId }: SessionOptions) {
    this.context = context
    this.apiDomain = apiDomain
    this._kmId = kmId
    this.apiKey = apiKey
  }

  get id(): string | undefined {
    return this._id
  }

  get kmId(): string {
    return this._kmId
  }

  private ensureId(): void {
    if (typeof this.id === 'undefined') {
      throw new Error('TODO: handle error')
    }
  }

  async start(): Promise<StartResponse> {
    const response = await Fetch.get<StartResponse>(
      `${this.apiDomain}/start/${this._kmId}`,
      { headers: { Authorization: `Basic ${btoa(`${this.apiKey}:`)}` } }
    )
    if (!response.result) {
      throw new Error('TODO: handle error')
    }

    this._id = response.result.id
    return response.result
  }

  async respond(answers: Answer[]): Promise<RespondResponse> {
    if (!this.id) {
      throw new Error('Session id is not defined. Call Session.Start first.')
    }

    const response = await Fetch.post<RespondResponse>(
      `${this.apiDomain}/${this.id}/respond`,
      { answers }
    )

    if (!response.result) {
      throw new Error('TODO: handle error')
    }

    // Set the kind property so that it is easier to distinguish the response types.
    if (typeof (response.result as ResultResponse).result !== 'undefined') {
      response.result.kind = 'result'
    } else if (
      typeof (response.result as QuestionResponse).question !== 'undefined'
    ) {
      response.result.kind = 'question'
    }

    return response.result
  }

  async inject(facts: Fact[]): Promise<InjectResponse> {
    if (!this.id) {
      throw new Error('Session id is not defined. Call Session.Start first.')
    }

    const response = await Fetch.post<InjectResponse>(
      `${this.apiDomain}/${this.id}/inject`,
      facts
    )

    if (!response.result) {
      throw new Error('TODO: handle error')
    }

    return response.result
  }
}

export interface SessionOptions {
  context?: string
  apiDomain: string
  kmId: string
  apiKey: string
}

export interface StartResponse {
  id: string
  kmVersion: {
    id: string
  }
}

export interface QuestionResponse {
  kind: 'question'
  question: {
    subject: string
    dataType: string // TODO: enum
    relationship: string
    type: string // TODO: enum
    plural: boolean
    allowCF: boolean
    canAdd: boolean
    prompt: string
    knownAnswers: [] // TODO: Type answers
    meta: {
      ruleStack: [{ name: string }]
    }
    sid: string
  }
}

export interface ResultResponse {
  kind: 'result'
  result: ResultFact[]
  stats: any // TODO: add types
  sid: string
}

export interface ResultFact extends Fact {
  objectMetadata?: string
  subjectMetadata?: string
  factID: string
}

export type RespondResponse = ResultResponse | QuestionResponse

export interface InjectResponse {
  result: 'OK'
}

export interface Answer extends Fact {
  // TODO: add props
}
