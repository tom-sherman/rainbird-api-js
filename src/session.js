const rb = require('./rainbird')

class Session {
  /**
   * Creates a new Session.
   */
  constructor (options) {
    if (typeof options === 'string') {
      this.apiKey = options
    } else if (options.apiKey) {
      this.apiKey = options.apiKey
    }
    if (!rb.validUuid(this.apiKey)) {
      throw new Error('Invalid API key')
    }

    /**
     * The API domain to make requests.
     * @type {string}
     */
    this.apiDomain = options.apiDomain || Session.DEFAULT_API_DOMAIN

    /**
     * The session ID returned from the /start method.
     * @type {?string}
     */
    this.id = null

    /**
     * The current Knowledge map ID.
     * @type {?string}
     */
    this.kmId = null

    /**
     * The current Knowledge Map version ID
     * @type {?string}
     */
    this.kmVersionId = null

    /**
     * A flag which determines if Session should store the previous responses.
     * @type {boolean}
     */
    this.recordHistory = options.recordHistory

    /**
     * An array of previous Rainbird responses.
     */
    this.history = []

    /**
     * The current query.
     * @type {{ subject: string, relationship: string, object: string }}
     */
    this.currentQuery = null
  }

  /**
   * Starts a new session.
   * Call without any arguments to get a new session with the same kmId and API options.
   * @param {string} [kmId] Knowledge map ID.
   */
  async start (kmId = this.kmId) {
    if (!rb.validUuid(kmId)) {
      throw new Error('Invalid Knowledge Map ID.')
    }

    this.kmId = kmId
    const session = await this.call({
      path: '/start/' + this.kmId,
      method: 'GET'
    })
    this.id = session.id
    // Clear history
    this.history.length = 0
    this.currentQuery = null
    return session
  }

  /**
   * Start a new query.
   * @param {object} options
   * @param {string} options.subject
   * @param {string} options.relationship
   * @param {string} options.object
   */
  async query ({ subject, relationship, object } = {}) {
    if (!this.id) {
      throw new Error('Session ID is not defined. Run Session.start before querying.')
    }
    const response = this.call({
      path: `/${this.id}/query`,
      body: { subject, relationship, object }
    })
    this.currentQuery = { subject, relationship, object }
    return response
  }

  /**
   * @typedef Fact
   * @property {string} subject
   * @property {string} relationship
   * @property {string} object
   * @property {number?} cf
   */

  /**
   * Respond to the previous question with one or more facts (answers).
   * @param {Fact[]} answers
   */
  async respond (answers) {
    if (!this.id) {
      throw new Error('Session ID is not defined. Run Session.start before querying.')
    }
    if (!Array.isArray(answers)) {
      answers = [ answers ]
    }
    for (const answer of answers) {
      if (!answer.cf) {
        answer.cf = 100
      }
    }
    return this.call({
      path: `/${this.id}/response`,
      body: { answers }
    })
  }

  inject (facts) {
    if (!this.id) {
      throw new Error('Session ID is not defined. Run Session.start before querying.')
    }
    if (!Array.isArray(facts)) {
      facts = [ facts ]
    }
    for (const fact of facts) {
      if (!fact.cf) {
        fact.cf = 100
      }
    }
    return this.call({
      path: `/${this.id}/inject`,
      body: facts
    })
  }

  /**
   * Sends a request to the Rainbird API.
   * @param {object} opts
   * @param {string} opts.path
   * @param {string} [opts.method = 'POST']
   * @param {object} [opts.qs]
   */
  async call (opts) {
    const response = await rb.call({
      ...opts,
      apiKey: this.apiKey,
      apiDomain: this.apiDomain
    })
    if (this.recordHistory) {
      this.history.push(response)
    }
    return response
  }

  static get DEFAULT_API_DOMAIN () {
    return rb.API_DOMAIN
  }
}

module.exports = Session
