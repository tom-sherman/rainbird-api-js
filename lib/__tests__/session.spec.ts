import { Session } from '../session'

it('can create new session', () => {
  expect(new Session({ apiKey: '', apiDomain: '', kmId: '' })).toMatchSnapshot()
})
