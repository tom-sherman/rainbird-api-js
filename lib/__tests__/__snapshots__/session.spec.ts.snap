// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can create new session 1`] = `
Session {
  "_kmId": "",
  "apiDomain": "",
  "apiKey": "",
  "context": undefined,
}
`;
